File: kotlin-validation/app/src/main/kotlin/oscarvarto/mx/domain/Person.kt
package oscarvarto.mx.domain

import arrow.core.Either
import arrow.core.NonEmptyList
import arrow.core.raise.either
import arrow.core.raise.ensure
import arrow.core.raise.zipOrAccumulate

sealed interface PersonValidationError
data object BlankName : PersonValidationError
data object NegativeAge : PersonValidationError
data object MaxAge : PersonValidationError

@ConsistentCopyVisibility
data class Person private constructor(val name: String, val age: Int) {
    companion object {
        const val MAX_AGE = 130

        operator fun invoke(
            name: String,
            age: Int,
        ): Either<NonEmptyList<PersonValidationError>, Person> = either {
            zipOrAccumulate(
                { ensure(name.isNotBlank()) { BlankName } },
                { ensure(age >= 0) { NegativeAge } },
                { ensure(age <= MAX_AGE) { MaxAge } },
            ) { _, _, _ -> Person(name, age) }
        }
    }
}

File: scala-validation/app/src/oscarvarto/mx/domain/Person.scala
package oscarvarto.mx.domain

import zio.prelude.Validation
import PersonValidationError.*

enum PersonValidationError:
  case BlankName, NegativeAge, MaxAge

final case class Person private (name: String, age: Int)

object Person:
  val MAX_AGE = 130

  def make(
      name: String,
      age: Int
  ): Validation[PersonValidationError, Person] =
    Validation.validateWith(
      if name.isBlank then Validation.fail(BlankName)
      else Validation.succeed(name),
      if age < 0 then Validation.fail(NegativeAge)
      else Validation.succeed(age),
      if age > MAX_AGE then Validation.fail(MaxAge)
      else Validation.succeed(age)
    )((n, a, _) => new Person(n, a))
